### PR TITLE
Add type and namespace declaration API

### PR DESCRIPTION
**Pull request by genlu**
This change was initially included in #34067. I have separated the change on compiler side and added unit tests.

new public compiler APIs (backed by declaration table)

```cs
public abstract partial class Compilation
{
    ...
    public abstract INamespaceDeclaration DeclarationRoot { get; }
    ...
}
```

```cs
namespace Microsoft.CodeAnalysis
{   
    public interface INamespaceOrTypeDeclaration
    {
        string Name { get; }

        ImmutableArray<INamespaceOrTypeDeclaration> Children { get; }

        bool IsNamespace { get; }

        bool IsType { get; }
    }

    public interface ITypeDeclaration : INamespaceOrTypeDeclaration
    {
        TypeKind TypeKind { get; }

        Accessibility DeclaredAccessibility { get; }

        int Arity { get; }

        new ImmutableArray<ITypeDeclaration> Children { get; }
    }

    public interface INamespaceDeclaration : INamespaceOrTypeDeclaration
    {
    }
}
```

@CyrusNajmabadi @jcouv @gafter 
FYI @dotnet/roslyn-compiler @dotnet/roslyn-ide

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 1]
```diff
@@ -3247,6 +3247,9 @@ internal bool EnableEnumArrayBlockInitialization
             }
         }
 
+        public override INamespaceDeclaration DeclarationRoot
+            => Declarations.GetDeclarationRoot(this);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: Possibly "GlobalDeclaration" given that we have Compilation.GlobalNamespace.  not too important, but since this is a public API we should try to land on what we think is best here.

File: src/Compilers/CSharp/Portable/Declarations/DeclarationTable.cs [hunk 2]
```diff
@@ -365,5 +366,8 @@ public IEnumerable<ReferenceDirective> ReferenceDirectives
 
             return false;
         }
+
+        public INamespaceDeclaration GetDeclarationRoot(CSharpCompilation compilation)
+            => GetMergedRoot(compilation);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Declarations/DeclarationTable.cs]: is this method actually needed?  couldn't this be inlined into the caller?

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 3]
```diff
@@ -3189,5 +3189,11 @@ public static string GetRequiredLanguageVersion(Diagnostic diagnostic)
 
             return foundVersion;
         }
+
+        /// <summary>
+        /// Gets The root namespace declaration that contains all namespaces and types declared in this compilation, 
+        /// merged into a single namespace hierarchy.
+        /// </summary>
+        public abstract INamespaceDeclaration DeclarationRoot { get; }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: see above about desired name here.

File: src/Compilers/Core/Portable/Declarations/INamespaceOrTypeDeclaration.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        /// <summary>
+        /// Returns true if this declaration is a namespace. If it is not a namespace, it must be a type.
+        /// </summary>
+        bool IsNamespace { get; }
+
+        /// <summary>
+        /// Returns true if this declaration is a type. If it is not a type, it must be a namespace.
+        /// </summary>
+        bool IsType { get; }
+    }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Declarations/INamespaceOrTypeDeclaration.cs]: while sufficient for the needs of completion, i really think we should also expose `ImmutableArray<Location> IDeclaration.NameLocations` for IDeclaration, and probably `ImmutableArray<SyntaxReference> SyntaxReferences` as well.  People are going to enumerate these and then ask the obvious question of "where is this, and how do i get to it?".

File: src/Compilers/Core/Portable/Declarations/ITypeDeclaration.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        TypeKind TypeKind { get; }
+
+        /// <summary>
+        /// Gets a <see cref="Accessibility"/> indicating the declared accessibility for this declaration.
+        /// Returns NotApplicable if no accessibility is declared.
+        Accessibility DeclaredAccessibility { get; }
+
+        /// <summary>
+        /// Returns the arity of this type declaration, or the number of type parameters it takes.
+        /// A non-generic type has zero arity.
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Declarations/ITypeDeclaration.cs]: you're missing end tags on several of these.

**Comment by CyrusNajmabadi** [comment]
Overall, the shape seems good to me.  just minor nits.  I'm sure compiler will have ideas aobut tests that are necessary.  Note: VB is interesting here because (i believe), they make declarations for things not in source.  So we'll likely need tests there.  @dotnet/roslyn-compiler can likely help you with a test plan here!

File: src/Compilers/Core/Portable/Declarations/INamespaceOrTypeDeclaration.cs (hunk 4 above)
Comment (genlu) [reply src/Compilers/Core/Portable/Declarations/INamespaceOrTypeDeclaration.cs]: Yes, you mentioned this in the other PR, and I agree with you that this is useful. But that can come later, I'd prefer to wait until some feature actually needs it.

File: src/Compilers/Core/Portable/Declarations/ITypeDeclaration.cs (hunk 5 above)
Comment (genlu) [reply src/Compilers/Core/Portable/Declarations/ITypeDeclaration.cs]: oops

File: src/Compilers/CSharp/Portable/Declarations/DeclarationTable.cs (hunk 2 above)
Comment (genlu) [reply src/Compilers/CSharp/Portable/Declarations/DeclarationTable.cs]: Sure. Will inline

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs (hunk 1 above)
Comment (genlu) [reply src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: Hmm, I'm not sure, global namespace is a thing whereas global declaration isn't. Thoughts? @dotnet/compiler-api

**Comment by genlu** [comment]
@CyrusNajmabadi One thing I'm aware of is the rootnamespace provided as an argument, I will add a test for that. Is there anything else I'm missing?

**Comment by CyrusNajmabadi** [comment]
> One thing I'm aware of is the rootnamespace provided as an argument,

Not sure what that means!

**Comment by genlu** [comment]
@CyrusNajmabadi I'm talking about this thing
http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/VisualBasicCompilationOptions.vb,362

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: it's a good question.  it might make sense to synthesize in that case?  i'm not actually sure what would be best :)

File: src/Compilers/Core/Portable/Declarations/INamespaceOrTypeDeclaration.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Declarations/INamespaceOrTypeDeclaration.cs]: I'll defer... but i personally think that's a mistake here.  As a public API, i'd prefer we not just make it for your single case :)  I'll let compiler decide what they think there.

**Comment by AlekseyTs** [comment]
@genlu Could you please provide details about the motivation for adding the APIs? What problem are we trying to solve?

**Comment by genlu** [comment]
@AlekseyTs The motivation for this API is the new IDE completion feature which provides types that are not in scope at the cursor location, but available in the compilation (either declared in the compilation or in a reference). (see the screenshot in https://github.com/dotnet/roslyn/pull/34067). When such an item is selected, we will add the corresponding using directive in addition to completing the type name. 

Basically, we will need an efficient way to get some syntactic info for all the top level types declared in the source (i.e. name, containing namespace, accessibility, type kind). This PR addresses this by exposing those information from declaration table.

**Comment by CyrusNajmabadi** [comment]
> Basically, we will need an efficient way

To dive into that a bit more.  Existing mechanisms for querying the global symbol table are somewhat unsatisfactory.  Currently, the only approaches would be:

1. using ISymbols for this.  However, that can involve a lot of churn as we will go realize every top-level symbol in a particular compilation.  Subsequent edits will cause this to happen over and over again.  We'd like to avoid that given the significant latency-goals of completion.
2. having the IDE side build an syntactic (i.e. file-based) index for all this information.  this is doable, but is another index to maintain, keep up to date, and persist.

The DeclTables solve the problem very neatly for us since they're already there, are already very efficient to maintain, and have been very stable over the lifetime of Roslyn.  As such, we're hoping we can elevate this internal implementation detail to a fully supported API which provides access to features like Completion which need that information about top-level symbols, but which would like to avoid actually forcing the symbols to actually be created.

Thanks!

**Comment by AlekseyTs** [comment]
@genlu, was this new API proposal presented to API Review Council? I am not sure whether this is a good idea to expose yet another way of looking at symbols vs. improving existing symbols API where it is needed.
CC @jaredpar, @gafter

**Comment by genlu** [comment]
@AlekseyTs No, this has not been reviewed by  API Review Council. Just discussion between @gafter @jcouv and myself.

**Comment by AlekseyTs** [comment]
> No, this has not been reviewed by API Review Council.

I think that before we spend time reviewing this PR, we should get an approval from the Council.

**Comment by genlu** [comment]
We ended up using ISymbol APIs for the new completion feature. Closed.
